instrumentation/apache-httpclient/apache-httpclient-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v4_0/ApacheHttpClientInstrumentationModule.java

```diff
@@ -147,127 +149,175 @@ public ApacheHttpClientInstrumentationModule() {
 
   public static class UriRequestAdvice {
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static SpanWithScope methodEnter(@Advice.Argument(0) HttpUriRequest request) {
-      int callDepth = CallDepthThreadLocalMap.incrementCallDepth(HttpClient.class);
-      if (callDepth > 0) {
-        return null;
+    public static void methodEnter(
+        @Advice.Argument(0) HttpUriRequest request,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") CallDepth callDepth) {
+      callDepth = tracer().getCallDepth();
+      if (callDepth.getAndIncrement() != 0) {
+        return;
+      }
+      Context parentContext = currentContext();
+      if (!tracer().shouldStartSpan(parentContext)) {
+        return;
       }
 
-      return ApacheHttpClientHelper.doMethodEnter(request);
+      context = tracer().startSpan(parentContext, request, request);
```

Comment:
I might be missing a detail, but instead of `shouldStartSpan` everywhere, can't `startSpan` just return `parentContext` if it shouldn't start a new span?

Reply:
the problem is how to know if the instrumentation should call end on the span in the returned context

Reply:
Ah remember now - had the idea a loong time ago (one of my first issues in the repo :D) that for this case, we might wrap the span in something that would still allow propagation and setting attributes (e.g., to enrich in HTTP instrumentation) without forwarding `end` or any other dangerous methods (update span name also probably not forwarded). If this seems like a reasonable idea not sure if it's better to do it now or later.

Reply:
I wrote up #1822 based on the discussion about this from yesterday